gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/oneconvergence_service_node_driver.py

```diff
@@ -0,0 +1,867 @@
+
+    def _update_router_routes(self, context, router_id, new_routes):
+        """
+        Adds extra routes to the router resource.
+
+        :param router_id: uuid of the router,
+        :param new_routes: list of new routes in this format
+                          "routes": [
+                                       {
+                                            "nexthop": "10.1.0.10",
+                                            "destination": "40.0.1.0/24"
+                                       },....
+                                    ]
+        """
+        admin_context = n_context.get_admin_context()
+        router_info = {"router": {"routes": new_routes}}
+        context.l3_plugin.update_router(admin_context, router_id, router_info)
+
+    def _create_port(self, context, admin_context, tenant_id, port_name,
+                     network_id, ip_address=None, subnet_id=None):
+        if subnet_id:
+            fixed_ips = [{"subnet_id": subnet_id}]
+        elif ip_address:
+            fixed_ips = [{"ip_address": ip_address}]
+        else:
+            # TODO(Magesh): Test if ATTR_NOT_SPECIFIED works fine
+            fixed_ips = attributes.ATTR_NOT_SPECIFIED
+
+        attrs = {'port': {'tenant_id': tenant_id,
+                          'name': port_name,
+                          'network_id': network_id,
+                          'fixed_ips': fixed_ips,
+                          'port_security_enabled': False,
+                          'admin_state_up': True,
+                          'mac_address': attributes.ATTR_NOT_SPECIFIED,
+                          'device_id': '',
+                          'device_owner': ''
+                          }
+                 }
+
+        try:
+            port = context.core_plugin.create_port(
+                admin_context, attrs)
+            self._notify_port_action(context, port, 'create')
+        except Exception:
+            LOG.exception(_("create port failed."))
+            return
+        return port
+
+    def _dhcp_agent_notifier(self, context):
+        # REVISIT(Magesh): Need initialization method after all
+        # plugins are loaded to grab and store notifier.
+        if not self._cached_agent_notifier:
+            agent_notifiers = getattr(
+                    context.core_plugin, 'agent_notifiers', {})
+            self._cached_agent_notifier = (
+                agent_notifiers.get(const.AGENT_TYPE_DHCP) or
+                dhcp_rpc_agent_api.DhcpAgentNotifyAPI())
+        return self._cached_agent_notifier
+
+    def _notify_port_action(self, context, port, action):
+        if cfg.CONF.dhcp_agent_notification:
+            self._dhcp_agent_notifier(context).notify(
+                context.plugin_context, {'port': port},
+                'port.' + action + '.end')
+
+
+## BASED OFF COMMIT 054a4fe14c4919771de71f4f24406b6c781efaef #########
+class OneConvergenceServiceNodeDriver(template_node_driver.TemplateNodeDriver):
```

Comment:
This needs to be updated to reflect Reference driver comments for renaming to HeatNodeDriver